File: app/src/main/java/com/lucas/omnia/activities/NewReplyActivity.java
package com.lucas.omnia.activities;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.MutableData;
import com.google.firebase.database.Transaction;
import com.google.firebase.database.ValueEventListener;
import com.lucas.omnia.R;
import com.lucas.omnia.databinding.ActivityNewReplyBinding;
import com.lucas.omnia.models.Comment;
import com.lucas.omnia.models.Reply;
import com.lucas.omnia.models.User;

import java.util.HashMap;
import java.util.Map;

import static com.lucas.omnia.activities.CommentsActivity.postKey;

public class NewReplyActivity extends BaseActivity {

    public static final String EXTRA_COMMENT_KEY = "comment_key";
    private static final String TAG = "NewReplyActivity";
    private static final String REQUIRED = "Required";

    private String commentKey;
    private DatabaseReference databaseReference;
    private DatabaseReference commentReference;
    private String author;

    private ActivityNewReplyBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityNewReplyBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        databaseReference = getDatabaseReference();

        // Get comment key from intent
        commentKey = getIntent().getStringExtra(EXTRA_COMMENT_KEY);
        if (commentKey == null) {
            throw new IllegalArgumentException("Must pass EXTRA_COMMENT_KEY");
        }

        commentReference = databaseReference.child("post-comments").child(postKey).child(commentKey);
        commentReference.addListenerForSingleValueEvent(
                new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        Comment comment = dataSnapshot.getValue(Comment.class);

                        author = comment.author;
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        Log.e(TAG, "loadReply:onCancelled", databaseError.toException());
                        setEditingEnabled(true);
                    }
                });

        binding.newReplyFabSubmit.setOnClickListener(v -> submitReply());
    }

    private void submitReply() {
        final String body = "@" + author + " " + binding.newReplyEtBody.getText().toString();
        final boolean edited = false;

        // Body is required
        if (TextUtils.isEmpty(body)) {
            binding.newReplyEtBody.setError(REQUIRED);
            return;
        }

        // Disable button so there are no multi-posts
        setEditingEnabled(false);
        Toast.makeText(this, "Replying...", Toast.LENGTH_SHORT).show();

        final String userId = getUid();
        databaseReference.child("users").child(userId).addListenerForSingleValueEvent(
                new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        User user = dataSnapshot.getValue(User.class);

                        if (user == null) {
                            Log.e(TAG, "User " + userId + " is unexpectedly null");
                            Toast.makeText(NewReplyActivity.this,
                                    getString(R.string.user_fetch_error),
                                    Toast.LENGTH_SHORT).show();
                        } else {
                            writeNewReply(userId, user.username, body, edited);
                            incrementRepliesCount();
                        }
                        setEditingEnabled(true);
                        finish();
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        Log.e(TAG, "getUser:onCancelled", databaseError.toException());
                        setEditingEnabled(true);
                    }
                });
    }

    private void setEditingEnabled(boolean enabled) {
        binding.newReplyEtBody.setEnabled(enabled);
        if (enabled) {
            binding.newReplyFabSubmit.show();
        } else {
            binding.newReplyFabSubmit.hide();
        }
    }

    private void writeNewReply(String userId, String username, String body, boolean edited) {
        // Create new item_reply at /comment-replies
        String key = databaseReference.child("comment-replies").child(commentKey).push().getKey();
        Reply reply = new Reply(userId, username, body, edited);
        Map<String, Object> replyValues = reply.toMap();

        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put("/comment-replies/" + commentKey + "/" + key, replyValues);
        databaseReference.updateChildren(childUpdates);
    }

    private void incrementRepliesCount() {
        commentReference.runTransaction(new Transaction.Handler() {
            @Override
            public Transaction.Result doTransaction(MutableData mutableData) {
                Comment c = mutableData.getValue(Comment.class);
                if (c == null) {
                    return Transaction.success(mutableData);
                }

                // Increment commentsCount of item_post
                c.replyCount = c.replyCount + 1;

                // Set value and report transaction success
                mutableData.setValue(c);
                return Transaction.success(mutableData);
            }

            @Override
            public void onComplete(DatabaseError databaseError, boolean committed,
                                   DataSnapshot currentData) {
                // Transaction completed
                Log.d(TAG, "ReplyTransaction:onComplete:" + databaseError);
            }
        });
    }
}

File: app/src/main/java/com/lucas/omnia/activities/NewReplyActivity.kt
package com.lucas.omnia.activities

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.widget.Toast
import com.google.firebase.database.*
import com.lucas.omnia.R
import com.lucas.omnia.databinding.ActivityNewReplyBinding
import com.lucas.omnia.models.Comment
import com.lucas.omnia.models.Reply
import com.lucas.omnia.models.User
import java.util.*

class NewReplyActivity : BaseActivity() {
    private var commentKey: String? = null
    private var databaseRef: DatabaseReference? = null
    private var commentReference: DatabaseReference? = null
    private var author: String? = null
    private var binding: ActivityNewReplyBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityNewReplyBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        databaseRef = databaseReference

        // Get comment key from intent
        commentKey = intent.getStringExtra(EXTRA_COMMENT_KEY)
        requireNotNull(commentKey) { "Must pass EXTRA_COMMENT_KEY" }
        commentReference = databaseRef!!.child("post-comments").child(CommentsActivity.postKey!!).child(commentKey!!)
        if (intent.hasExtra(EXTRA_REPLY_KEY)) {
            // Get reply key from intent
            val replyKey = intent.getStringExtra(EXTRA_REPLY_KEY)
                    ?: throw IllegalArgumentException("Must pass EXTRA_REPLY_KEY")
            val replyReference = databaseRef!!.child("comment-replies").child(commentKey!!).child(replyKey)
            replyReference.addListenerForSingleValueEvent(
                    object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val reply = dataSnapshot.getValue(Reply::class.java)
                            author = reply!!.author
                        }

                        override fun onCancelled(databaseError: DatabaseError) {
                            Log.e(TAG, "loadReply:onCancelled", databaseError.toException())
                            setEditingEnabled(true)
                        }
                    })
        } else {
            commentReference!!.addListenerForSingleValueEvent(
                    object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val comment = dataSnapshot.getValue(Comment::class.java)
                            author = comment!!.author
                        }

                        override fun onCancelled(databaseError: DatabaseError) {
                            Log.e(TAG, "loadReply:onCancelled", databaseError.toException())
                            setEditingEnabled(true)
                        }
                    })
        }
        binding!!.newReplyFabSubmit.setOnClickListener { submitReply() }
    }

    private fun submitReply() {
        val body = "@" + author + " " + binding!!.newReplyEtBody.text.toString()

        // Body is required
        if (TextUtils.isEmpty(body)) {
            binding!!.newReplyEtBody.error = REQUIRED
            return
        }

        // Disable button so there are no multi-posts
        setEditingEnabled(false)
        Toast.makeText(this, getString(R.string.new_reply_toast_replying), Toast.LENGTH_SHORT).show()
        val userId = uid
        databaseRef!!.child("users").child(userId).addListenerForSingleValueEvent(
                object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val user = dataSnapshot.getValue(User::class.java)
                        if (user == null) {
                            Log.e(TAG, "User $userId is unexpectedly null")
                            Toast.makeText(this@NewReplyActivity,
                                    getString(R.string.new_post_toast_user_fetch_error),
                                    Toast.LENGTH_SHORT).show()
                        } else {
                            writeNewReply(userId, user.username, body)
                            incrementRepliesCount()
                        }
                        setEditingEnabled(true)
                        finish()
                    }

                    override fun onCancelled(databaseError: DatabaseError) {
                        Log.e(TAG, "getUser:onCancelled", databaseError.toException())
                        setEditingEnabled(true)
                    }
                })
    }

    private fun setEditingEnabled(enabled: Boolean) {
        binding!!.newReplyEtBody.isEnabled = enabled
        if (enabled) {
            binding!!.newReplyFabSubmit.show()
        } else {
            binding!!.newReplyFabSubmit.hide()
        }
    }

    private fun writeNewReply(userId: String, username: String?, body: String) {
        // Create new item_reply at /comment-replies
        val key = databaseRef!!.child("comment-replies").child(commentKey!!).push().key
        val reply = Reply(userId, username, body)
        val replyValues = reply.toMap()
        val childUpdates: MutableMap<String, Any> = HashMap()
        childUpdates["/comment-replies/$commentKey/$key"] = replyValues
        databaseRef!!.updateChildren(childUpdates)

        // Set timestamp
        databaseRef!!.child("comment-replies").child(commentKey!!).child(key!!).child("timestamp").setValue(ServerValue.TIMESTAMP)
    }

    private fun incrementRepliesCount() {
        commentReference!!.runTransaction(object : Transaction.Handler {
            override fun doTransaction(mutableData: MutableData): Transaction.Result {
                val c = mutableData.getValue(Comment::class.java)
                        ?: return Transaction.success(mutableData)

                // Increment commentsCount of item_post
                c.replyCount = c.replyCount + 1

                // Set value and report transaction success
                mutableData.value = c
                return Transaction.success(mutableData)
            }

            override fun onComplete(databaseError: DatabaseError?, committed: Boolean,
                                    currentData: DataSnapshot?) {
                // Transaction completed
                Log.d(TAG, "ReplyTransaction:onComplete:$databaseError")
            }
        })
    }

    companion object {
        const val EXTRA_COMMENT_KEY = "comment_key"
        const val EXTRA_REPLY_KEY = "reply_key"
        private const val TAG = "NewReplyActivity"
        private const val REQUIRED = "Required"
    }
}